### PR TITLE
[ntuple] fix setting field IDs for complex fields in RDF DS

### LINKED ISSUE #12852: [ntuple] RDataFrame `Display` crashes on custom classes
- [x] Checked for duplicates

### Describe the bug
Calling `Display` on an RDataFrame created from an RNTuple that contains user-defined classes causes ROOT to crash with an error message `No on-disk column information for for field X`, where `X` is a data member of the class in question.

Other methods appear to be working correctly, as well as explicitly specifying which data member should be read (i.e. `df.Display({"electrons.pt"})`).

### Expected behavior
Nothing crashes and an `RDisplay` object is returned.

### To Reproduce
Minimal reproducer:
```c++
#include <ROOT/RDataFrame.hxx>
#include <ROOT/RNTuple.hxx>
#include <ROOT/RNTupleDS.hxx>
#include <ROOT/RNTupleModel.hxx>

#include <iostream>

using ROOT::Experimental::RNTupleModel;
using ROOT::Experimental::RNTupleWriteOptions;
using ROOT::Experimental::RNTupleWriter;

const std::string kNTupleName = "N";
const std::string kNTupleFileName = "ntuple_electrons.root";

struct Electron {
  float pt = 0;
};

void write() {
  auto model = RNTupleModel::Create();
  auto fldElectrons = model->MakeField<Electron>("electrons");

  auto ntuple = RNTupleWriter::Recreate(std::move(model), kNTupleName, kNTupleFileName);

  fldElectrons->pt = 42.;
  ntuple->Fill();
}

void read() {
  auto df = ROOT::RDF::Experimental::FromRNTuple(kNTupleName, kNTupleFileName);
  auto dfDef = df.Define("ptSquared", "electrons.pt * electrons.pt");

  dfDef.Describe().Print();

  auto dfDisp = df.Display();
  dfDisp->Print();
}

void ntuple_rdf_custom_class() {
  write();
  read();
}
``` 

### Setup
1. ROOT version: master
2. Operating system: Ubuntu 22

### PR DESCRIPTION
**Pull request by jblomer**
# This Pull request:

Fix setting the field IDs of fields with subfields in the RNTuple RDF data source. It also fixes field IDs for clones of RClassField instances.

## Checklist:

- [X] tested changes locally

This PR fixes #12852

### CODE REVIEW

File: tree/ntuple/v7/src/RField.cxx [hunk 1]
```diff
@@ -1219,7 +1245,9 @@ void ROOT::Experimental::RClassField::AddReadCallbacksFromIORules(const std::spa
 std::unique_ptr<ROOT::Experimental::Detail::RFieldBase>
 ROOT::Experimental::RClassField::CloneImpl(std::string_view newName) const
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:1199]: Good catch, thank you!

File: tree/ntuple/v7/src/RField.cxx [hunk 2]
```diff
@@ -206,6 +206,32 @@ std::tuple<const void *const *, const std::int32_t *, const std::int32_t *> GetR
...
+      fromSubfields.pop_front();
+      auto sTo = toSubfields.front();
+      toSubfields.pop_front();
+      sTo->SetOnDiskId(sFrom->GetOnDiskId());
+
+      for (const auto f : sFrom->GetSubFields())
+         fromSubfields.emplace_back(f);
+      for (auto f : sTo->GetSubFields())
+         toSubfields.emplace_back(f);
+   }
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx]: Yep, that's a lot nicer! I needed to add const schema iterators along the way.

File: tree/ntuple/v7/src/RField.cxx [hunk 3]
```diff
@@ -207,6 +207,18 @@ std::tuple<const void *const *, const std::int32_t *, const std::int32_t *> GetR
...
+/// Applies the field IDs from 'from' to 'to', where from and to are expected to be each other's clones.
+/// Used in RClassField and RCollectionClassField cloning. In these classes, we don't clone the subfields
+/// but we recreate them. Therefore, the on-disk IDs need to be fixed up.
+void SyncFieldIDs(const ROOT::Experimental::Detail::RFieldBase &from, ROOT::Experimental::Detail::RFieldBase &to)
+{
+   auto iFrom = from.cbegin();
+   auto iTo = to.begin();
+   for (; iFrom != from.cend(); ++iFrom, ++iTo) {
+      iTo->SetOnDiskId(iFrom->GetOnDiskId());
+   }
```
Comment (jalopezg-git) [inline tree/ntuple/v7/src/RField.cxx:219]: Compared to the previous version in this PR, the `RSchemaIteratorTemplate<bool>` only visits descendant fields.  Instead, I think it should be
```suggestion
   to.SetOnDiskId(from.GetOnDiskId());
   auto iTo = to.begin();
   for (auto iFrom = from.cbegin(); iFrom != from.cend(); ++iFrom, ++iTo) {
      iTo->SetOnDiskId(iFrom->GetOnDiskId());
   }
```

Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:219]: It's true that we only visit descendant fields but that should be enough. The ID of the mother field itself is set by `Clone()`.

**Review by jalopezg-git — APPROVED** [review]
LGTM :+1:!

File: tree/ntuple/v7/src/RField.cxx (hunk 3 above)
Comment (jalopezg-git) [reply tree/ntuple/v7/src/RField.cxx:219]: Oh, that's absolutely right!

**Comment by jalopezg-git** [comment]
One last (post-merge) comment: given that `SyncFieldIDs()` modifies one of the arguments, we could perhaps turn it into a private member function of `RFieldBase`.  I don't have a strong opinion though.

**Pull request merged**